controllers/helpers.go

```diff
@@ -272,6 +272,7 @@ func newCloudProviderConfig(d azure.ClusterScoper) (controlPlaneConfig *CloudPro
 			SubnetName:                   subnet.Name,
 			RouteTableName:               subnet.RouteTable.Name,
 			LoadBalancerSku:              "Standard",
+			LoadBalancerName:             d.OutboundLBName(infrav1.Node),
```

Comment:
Do you know if this impacts/applies to the internal LB as well? I know that cloud-provider uses `clusterName` for public LB and `clusterName-internal` for internal LB by default today

Reply:
I will look into it and test it.

Reply:
Ok. So this does not apply to loadbalancers created via capz in general. So we are good. 
For private cluster, the apiserver lb name is : `<cluster-name>-internal-lb`
For public cluster, the apiserver lb name is : `<cluster-name>-public-lb`
And this PR, makes sure that whenever a LB is created for Services of type loadblancer it uses the same name as that of worker node outbound lb otherwise defaults to <clustername>. 
So we are good.

Reply:
what about the internal LBs created by cloud-provider for private services?

Reply:
>what about the internal LBs created by cloud-provider for private services?
Trying to understand it. What are private services?
For k8s services we have ExternalName, ClusterIP, NodePort, and LoadBalancer only.

Reply:
it's a service of type LoadBalancer with the annotation `service.beta.kubernetes.io/azure-load-balancer-internal` 

as demonstrated by this e2e test https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_lb.go#L120 at https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b232461616673714a8255d74e6d77a0068f8e00f/test/e2e/kubernetes/deployment/deployment.go#L186

Reply:
I see. Let me look at that again and I will get back.

Reply:
It does impact the internal LB naming as well. So the name is <loadbalancer-name-in-azure.json>-internal. And I do not see a problem/ backward compatibility issue.

The only internal type LB that capz creates is for apiServerLB type of `Internal` for private cluster whose name is <cluster-name>-internal-lb. And in case when a service of internal type loadbalancer is created, the name will be <loadbalancer-name-in-azure.json>-internal and both have two different backend pools to target so it should be fine. 

-----------

For existing private clusters with service of internal lb type -- though the name of the lb will be <clustername>-internal and with this new change the name should be <loadbalancer-name-in-azure.json>-internal. Both the names will not be same if a outboundLB is specified. And this may need some upgrade handling.  
I tested this by changing the loadbalancer name in azure.json and restarting the kube-controller-manager and found out that, the existing internal lb service existed with the old name and nothing happened to that but the new service that I created caused creation of a new LB with a different name but has not active backend pools and this was expected. 

But the surprising thing was -- both the services are routable from within the subnet, probably because the frontend IP comes from the same subnet. ( Though this looks a bit odd to me )

Anyways, to have the loadbalancers right, deleting an re creating the service should fix it after the upgrade.

cc @andyzhangx @CecileRobertMichon

Reply:
>But the surprising thing was -- both the services are routable from within the subnet, probably because the frontend IP comes from the same subnet. ( Though this looks a bit odd to me )


Oh! I understand this, because I was CURLing from the node. Thanks to @jayunit100

Reply:
Is there an edge case that needs to be handled if there is _no_ outbound LB specified?

Reply:
looks like we return `""` if that's the case which should let cloud-provider-azure default the `LoadBalancerName` as before

Reply:
> So the name is <loadbalancer-name-in-azure.json>-internal. And I do not see a problem/ backward compatibility issue.

that's the main thing I was looking for

> For existing private clusters with service of internal lb type

for private clusters, only the API Server LB is private. This should not affect the service LB as it only has the control planes in its backend pools. So CAPZ right now never creates an internal LB with the worker nodes in the backend AFAIK, which means we're good.

The only thing we have to ensure is that the previous cloud-provider default for private LB name is the same as the new one otherwise it could cause issues for existing services. I think the only backcompat issue that could arise here is if someone is using a custom node outbound LB name that is different from cluster name, but then I think this setup would already be broken today.

Reply:
@jackfrancis to verify my assumptions

Reply:
There's a lot to chew on here, but nothing jumped out as wrong. Are there any obvious add'l cluster scenarios we can add to our tests that would have given us a more data-driven answer to many of these back-compat questions?

Reply:
Our default e2e test includes both an external and internal LB Service test, and our optional test covers the private cluster scenario, I triggered that yesterday and it [passed](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/3059/pull-cluster-api-provider-azure-e2e-optional/1620908010131951616) (however not that we're not testing LBs on that, we would need #2963 to get it done, I don't think it's worth holding this bug fix PR for it since @sonasingh46 ran that validation manually)

Reply:
I will take a shot on #2963 and raise a PR next week for more confidence but I think this can get merged if there are no objections. Happy to answer any question that folks still have.